### PR TITLE
feat: Include file path option for NDJSON

### PR DESCRIPTION
**Pull request by nameexhaustion**
ref https://github.com/pola-rs/polars/issues/10481

### CODE REVIEW

File: crates/polars-mem-engine/src/executors/scan/ndjson.rs [hunk 1]
```diff
@@ -40,6 +40,18 @@ impl JsonExec {
 
         let mut n_rows = self.file_scan_options.n_rows;
 
+        // Avoid panicking
+        if n_rows == Some(0) {
```
Comment (nameexhaustion) [inline crates/polars-mem-engine/src/executors/scan/ndjson.rs:44]: drive-by - fix panic on `scan_ndjson().head(0)`

File: crates/polars-plan/src/plans/conversion/dsl_to_ir.rs [hunk 2]
```diff
@@ -185,6 +185,19 @@ pub fn to_alp_impl(
...
+                file_options.include_file_paths.filter(|_| match scan_type {
+                    #[cfg(feature = "parquet")]
+                    FileScan::Parquet { .. } => true,
+                    #[cfg(feature = "ipc")]
+                    FileScan::Ipc { .. } => true,
+                    #[cfg(feature = "csv")]
+                    FileScan::Csv { .. } => true,
+                    #[cfg(feature = "json")]
+                    FileScan::NDJson { .. } => true,
+                    FileScan::Anonymous { .. } => false,
```
Comment (nameexhaustion) [inline crates/polars-plan/src/plans/conversion/dsl_to_ir.rs:198]: make sure the lazy schema is correct

File: py-polars/polars/io/ndjson.py [hunk 3]
```diff
@@ -138,12 +139,16 @@ def scan_ndjson(
         Offset to start the row index column (only use if the name is set)
     ignore_errors
         Return `Null` if parsing fails because of schema mismatches.
+    include_file_paths
+        Include the path of the source file(s) as a column with this name.
     """
     if isinstance(source, (str, Path)):
-        source = normalize_filepath(source)
+        source = normalize_filepath(source, check_not_directory=False)
```
Comment (nameexhaustion) [inline py-polars/polars/io/ndjson.py:146]: drive-by - enable scanning directories

File: crates/polars-mem-engine/src/executors/scan/ndjson.rs (hunk 1 above)
Comment (ritchie46) [reply crates/polars-mem-engine/src/executors/scan/ndjson.rs:44]: Can you add a test for this one?

Comment (nameexhaustion) [reply crates/polars-mem-engine/src/executors/scan/ndjson.rs:44]: added a test below
